### DISCUSSION TITLE
"Positions ... are out of range ..." panic on `:wa` of an empty file from the another view

### DISCUSSION THREAD

**Discussion by nuid64** [helix-editor/helix#9157] [Troubleshooting]
### Summary

Important for view of the empty file to be left or up of the view active at the moment buffers are written.

### Reproduction Steps

I tried this:

1. `hx f b`
2. `<space>ws` / `<space>/wv`
3. `<space>bb<enter>` (pick buffer b)
4. `<space>wk` / `<space>wh`
5. `d` (leaving the file empty)
6. `<space>wj` / `<space>wl`
7. `:wa`

I expected this to happen:
Add blank line.

Instead, this happened:
```
thread 'main' panicked at helix-core/src/transaction.rs:461:9:
Positions [(1, Before)] are out of range for changeset len 0!
```

### Helix log

Nothing useful

### Platform

Linux 6.6.7

### Terminal Emulator

Alacritty 0.12.3

### Installation Method

source

### Helix Version

23.10-146-gd5e6749f

**Comment by TornaxO7** [comment]
Hm... I'm not able to reproduce this with the master branch. Could you please try it with the master branch?

**Reply by nuid64** [reply]
I built Helix just yesterday with the master branch, it is up to date

**Reply by TornaxO7** [reply]
Hm... may I ask if you can share your config?

**Reply by gabydd** [reply]
I think this is a probably a problem with insert-final newline see the backtrace:
<details><summary>Relevant backtrace</summary>
<p>


```
thread 'main' panicked at 'Positions [(1, Before)] are out of range for changeset len 0!', helix-core/src/transaction.rs:461:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:578:5
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/core/src/panicking.rs:67:14
   2: helix_core::transaction::ChangeSet::update_positions
             at ./helix-core/src/transaction.rs:461:9
   3: helix_core::selection::Selection::map_no_normalize
             at ./helix-core/src/selection.rs:483:9
   4: helix_core::selection::Selection::map
             at ./helix-core/src/selection.rs:455:9
   5: helix_view::view::JumpList::apply
             at ./helix-view/src/view.rs:95:30
   6: helix_view::view::View::apply
             at ./helix-view/src/view.rs:599:9
   7: helix_view::document::Document::append_changes_to_history
             at ./helix-view/src/document.rs:1451:9
   8: helix_term::commands::typed::insert_final_newline
             at ./helix-term/src/commands/typed.rs:373:9
   9: helix_term::commands::typed::write_all_impl
             at ./helix-term/src/commands/typed.rs:718:13
  10: helix_term::commands::typed::write_all
             at ./helix-term/src/commands/typed.rs:757:5
  11: helix_term::commands::typed::command_mode::{{closure}}
             at ./helix-term/src/commands/typed.rs:3174:33
  12: <alloc::boxed::Box<F,A> as core::ops::function::FnMut<Args>>::call_mut
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/alloc/src/boxed.rs:1980:9
  13: <helix_term::ui::prompt::Prompt as helix_term::compositor::Component>::handle_event
             at ./helix-term/src/ui/prompt.rs:595:21
  14: helix_term::compositor::Compositor::handle_event
             at ./helix-term/src/compositor.rs:152:19
  15: helix_term::application::Application::handle_terminal_events::{{closure}}
             at ./helix-term/src/application.rs:646:22
  16: helix_term::application::Application::event_loop_until_idle::{{closure}}
             at ./helix-term/src/application.rs:325:55
  17: helix_term::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:299:57
  18: helix_term::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:1176:38
  19: hx::main_impl::{{closure}}
             at ./helix-term/src/main.rs:161:53
```


</p>
</details>

**Reply by gabydd** [reply]
one thing that fixes this which I am not sure is the correct fix (cc: @the-mikedavis or @pascalkuthe ) is to add:
```rs
let view = view_mut!(cx.editor, target_view);
view.sync_changes(doc);
 ```
 in ``write_all_impl`` just before the call to ``insert_final_newline`` this might also fix some of the other crashes with write-all because what it looks like to me is that write_all picks an arbitrary view, but that view might not have all the transactions applied and them, and then presumably there is an invalid selection or something along those lines that gets mapped through adding the newline or some formatting, which is true in this case because it picks the view that buffer b is open in which doesn't have the deletion applied.

**Reply by gabydd** [reply]
ah pascal already figured this out a while back: https://github.com/helix-editor/helix/issues/6752#issuecomment-1508985899

**Reply by pascalkuthe** [reply]
Ah good catch, I couldnt reproduce but this is probably a bit tricky. 
This is a duplicate of that issue then. It shouldn't be too hard to fix writeall your fixes is mostly correct. Please create a PR! You should only called syncview if the view is not the active view tough (a simple != suffices).


This bug exists in multiple other places in the codebase we basically need to do this anywhere where we append to history in the non-active view. It may be worth creating a unity function for this and inserting calls everywhere this could comeup
